Finalize Maze Generation

jbdyn opened the pull request:
Closes #945

jbdyn committed:
Remove partition index counter from binary space partitioning routine

Use list `pop` and `append` methods instead

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -228,22 +218,22 @@ def add_wall_and_split(partition, walls, ngaps, vertical, rng=None):
-
-        # increase the partition index
-        p += 1
+        partitions.extend(new[::-1])
```
That's what I am used to: from left to right and from top to bottom. Helps me to visualize things without too much thinking. But sure, technically I could just revert the list entries.

... However, I am still not entirely happy with the `if vertical` cases. Maybe I see a way to get rid of them without sacrificing readability and without introducing more complexity.

jbdyn committed:
Define `add_wall_and_split` as pure function

jbdyn commented:
A quick summary of what I did or tried to do:

- tame the binary space partitioning by distinguishing between partition data and partition usage: introduce a new class `Partition`, simplify the partitioning loop, make out partitioning parameters regarding `Partition.length` and `ngaps`
- define the region of a `Partition` such that its surrounding walls are included to make splitting and wall sampling a bit easier to understand
- keep the familiar Pelita-Maze-Look while breaking compatibility with the NumPy version by using `rng.sample` for border and inner wall sampling instead of `rng.shuffle`

Upcoming:
- adapt tests

Unfortunatly, I was not able to get away with a single `if vertical: ... else ...` block since sorting partitions out and wall sampling need to run independently, so I had at least two of those blocks. And the could was hard to parse, so I decided to make a 180° turn and put readability and simplification over code length and speed (the code is slower by ~25%).

Despite that, I am especially excited about the partitioning parameters. With them we could adjust the mazes foreseeably.

Debilski reviewed (commented):
Thanks for the work. I haven’t looked at it in too much detail but I’ve added added a few comments.

A few more general remarks: The `Partition` class looks slightly over-engineered, but if it allows us to expose some partition properties in an invariant way, then I guess it makes sense. :)

I think I would move all logic to the split function and then remove the need of having to use a rng in `__init__`. (Or the other way round: We could already do the whole splitting in `__init__` as well and keep the sub-partitions in a list and evaluate the whole tree …)

Some further design ideas regarding the algorithm (these do not need to be implemented now but I think some ideas are useful to keep in mind while writing the code):

- Currently we split horizontal, vertical, horizontal, … I think we could also split in an irregular pattern using the rng (with some bias maybe)
- There is no reason for splitting in halves or is there? So we might also have more than two sub-partitions (again perhaps drawn with a biased rng). (Maybe irregular splitting is also equivalent to splitting in more than two sub-partitions?)

jbdyn committed:
Restore NumPy compatibility for mazes with odd height

jbdyn committed:
Redefine a partition to include its surrounding walls

Instead of being the area inside these walls

jbdyn committed:
Introduce temporary API wrapper for sampling walls

The sampling still shuffles candidates for removal instead of sampling the candidates to keep directly

jbdyn committed:
Use sets to filter out gaps from inner walls

jbdyn committed:
Rework definitions for maze variables

jbdyn committed:
Take maze partition measurements in the `u`-`v`-space

In `u`-`v`-space, the inner partition wall is always vertical, which means that the wall position is always sampled on the `u` axis and the wall length is measured along the `v`-axis

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -118,185 +118,197 @@ def distribute_food(all_tiles, chamber_tiles, trapped_food, total_food, rng=None
-                wall.discard((pos, ymin + gap))
+            return [(x, self.pos) for x in range(self.xmin, self.xmax + 1)]
+
+    def split(self):
```
Yes, this would have been better. See also https://github.com/ASPP/pelita/pull/946#discussion_r2550368668

jbdyn commented:
In [0ac6d44](https://github.com/ASPP/pelita/pull/946/commits/0ac6d4431c99e658383345b5439309559643c5f9), I got rid of all the `if vertical ... else ...` cases and recurring code doubling.

Also, I think my obsession with variable naming payed off here, clearifying variable relations and aligning similar lines of code.

(I hope you like it :smirk:)

jbdyn committed:
Update comments for the binary space partitioning algorithm

jbdyn committed:
Update comments for the generation of the left half of the maze

jbdyn committed:
Substract the bot positions from sampled walls

jbdyn committed:
Rename maze space transformations

jbdyn committed:
Add test for finding chambers in a half maze

jbdyn committed:
Make `walls_to_graph` handle open mazes

jbdyn committed:
Extend partition documentation

jbdyn committed:
Restrict chamber finding on the left maze half

Connections between border gap pairs, which would be connected after mirroring, emulate the continuation on the right maze

jbdyn committed:
Add comments for module constants

jbdyn committed:
Rename border-related variables

jbdyn committed:
Rename `add_wall_and_split` to `add_inner_walls` and slightly change its API

jbdyn committed:
Link the return value signature of `find_trapped_tiles` to keyword argument

jbdyn committed:
Redefine `mirror` to not unite original and mirrored sets

otizonaizit commented:
Hey, yes, I am going to review this very soon. In the meantime could you have a look at this particular case here? Look at the thing in the middle in this screenshot:

```bash
pelita --seed 3582404332638874406

```
<img width="902" height="632" alt="screenshot1" src="https://github.com/user-attachments/assets/a279a886-7391-423e-82b1-879f6bf666bd" />


So, with the current implementation of `find_trapped_tails`, but it is the same in `v2.6.0`, so this hasn't changed, the thing in the middle is freaking up our algorithm. Proof:

```bash
# no food in chambers
pelita --seed 3582404332638874406 --food 0:30
```
<img width="902" height="567" alt="screenshot2" src="https://github.com/user-attachments/assets/fc942948-3a0f-44a4-89f9-4576fa9c1c4e" />

```bash
# as much food as possible in chambers
pelita --seed 3582404332638874406 --food 30:30
```
<img width="902" height="567" alt="screenshot4" src="https://github.com/user-attachments/assets/ca5527fe-027a-4cef-9023-62e270c5947a" />


which means that the only part of the maze which is considered not a chamber is the 8 free tiles at the border near the line with 8 food pellets. Is this in line with what we define a chamber? Or is this an artifact of the algorithm? To me it seems that the intuitive definition of chamber is a place that has a entrance of 1, the problem is: from which side are you entering? In this case, if you enter from the right, then the chamber is the whole maze except the thing in the middle, if you enter from the left is the opposite.

What do you think should we do in this case? Is it worth reasoning about this or should we just deem this seed as a crazy one and move on with life?

jbdyn commented:
@otizonaizit That looks weird, but it's indeed in line with our definition of chambers :grin:
I marked the chamber tiles for this case with dots:

```
################################
#..............#               #
#..............#               #
##.#.####.#.####               #
#..#.#.......#                 #
#....#...#...#                 #
##.#.#.....#.#                 #
#..#.#...#...#                 #
#..#.###.#.#..                 #
#..#.....#.#.#                 #
#....##.######                 #
###.##.......#                 #
#....##.########               #
#.##.#.........#               #
#............#.#               #
################################
```

The chambers we are interested in are actually the "side" ones which have a single articulation point in common with the "main" chamber.
So, for 3582404332638874406, we have one big "side" chamber on the left, as there is only a single one-tile-big entrance to it.

We can thereby go on with life - or reason about making the total food count dependent on the chamber size, i.e. giving the number of food pellets in terms of percentages instead of absolute numbers. This would avoid such surprises, but actually I like these kind of quirky cases.
I am fine to leave it as is. :relaxed:

otizonaizit commented on pelita/maze_generator.py:
```diff
@@ -55,21 +63,17 @@ def sample_nodes(nodes, k, rng=None):
 
 
-def find_trapped_tiles(graph, width, include_chambers=False):
+def find_trapped_tiles(graph, gaps, include_chambers=False):
```
I think that `gaps` is not a good name for the argument, or we need a doc-string that explains what it means. In general, maybe this could be by default `None`, so that `find_chambers` can be tested and used independently of our particular use case? Not sure if the generalization is useful, but still thinking about us in 5 years looking back, I think it will be odd if we don't explain this or make it general.

jbdyn replied:
I made that one clearer by renaming `gaps` to `border_gaps` in f4073a0a.
The `include_chambers` argument has been removed in bfbb5108 in favor of the implementation of its corresponding logic in the maze generation test module.

otizonaizit commented on pelita/maze_generator.py:
```diff
@@ -55,21 +63,17 @@ def sample_nodes(nodes, k, rng=None):
-            # only the main chamber covers both sides
-            # our own mazes should only have one central chamber
-            # but other configurations could have more than one
+        if (chamber & gaps):
```
I am not sure I get the logic. It clearly works, so I just need help to understand why. So, we have the following maze:
```
####################
#   .....XX        #
#   #####XX        #
#        XX        #
#        XX        #
#        XX        #
#        XX#####   #
#        XX        #
####################
```

the tiles with the dots are within a chamber, are they not? And this chamber intersects the border (denoted by `X`), right? So how can this be automatically assigned to the main chamber?

otizonaizit commented on pelita/maze_generator.py:
```diff
@@ -252,53 +289,80 @@ def generate_half_maze(width, height, ngaps_center, bots_pos, rng=None):
+        gaps.add(lower)
+
+        # edges between those gaps which would be connected after mirroring
+        edges.add((upper, lower))
```
I like this trick, but I am afraid this will be very confusing in the long run. Maybe it is enough to find a better name for this? After all these are "fake" edges, in the sense that they are not going to be there once we have the full maze, right? It should be clear that this here is a trick.

jbdyn replied:
I renamed the `edges` to `border_bridges` and rephrased the corresponding comments in 98339855.

otizonaizit commented on pelita/team.py:
```diff
@@ -82,10 +82,9 @@ def walls_to_graph(walls, shape=None):
-                    if neighbor not in walls:
+                    # check if the new node is free and within the bounds as
+                    # the walls might not be closed
+                    if neighbor not in walls and neighbor[0] < width and neighbor[1] < height:
```
I guess this could be changed so that the `if` exits faster when one condition is false? Wouldn't this be the same logic?
```python
if neighbor in walls or neighbor[0] >= width or neighbor[1] >= height:
    continue
```
Given that this function is one of the bottlenecks in maze generation but also, unfortunately, also for `Bot`, because the graph gets generated once for each `Bot` instance, it would be cool to optimize as much as possible. I am not sure my suggestion above does indeed improve things, but I guess it should?
```

Debilski replied:
In the game it should only run once per team (and otherwise be cached), so I think it is ok. Also the function spends 60% of the time in `graph.add_edge` which we don’t have any control over. At best, we can therefore half the time.

jbdyn replied:
Then we leave it as is for now.

otizonaizit reviewed (changes requested):
In general I am really liking this. It is work well invested, I am sure our future selves are going to be for ever grateful that this has been redesigned with legibility in mind :-)
I left some minor/middle-sized comments.

otizonaizit commented:
> @otizonaizit That looks weird, but it's indeed in line with our definition of chambers 😁 I marked the chamber tiles for this case with dots:
> 
> ```
> ################################
> #..............#               #
> #..............#               #
> ##.#.####.#.####               #
> #..#.#.......#                 #
> #....#...#...#                 #
> ##.#.#.....#.#                 #
> #..#.#...#...#                 #
> #..#.###.#.#..                 #
> #..#.....#.#.#                 #
> #....##.######                 #
> ###.##.......#                 #
> #....##.########               #
> #.##.#.........#               #
> #............#.#               #
> ################################
> ```
> 
> The chambers we are interested in are actually the "side" ones which have a single articulation point in common with the "main" chamber. So, for 3582404332638874406, we have one big "side" chamber on the left, as there is only a single one-tile-big entrance to it.
> 
> We can thereby go on with life - or reason about making the total food count dependent on the chamber size, i.e. giving the number of food pellets in terms of percentages instead of absolute numbers. This would avoid such surprises, but actually I like these kind of quirky cases. I am fine to leave it as is. ☺️

OK, so let's go on with our lives :-)

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -118,132 +122,165 @@ def distribute_food(all_tiles, chamber_tiles, trapped_food, total_food, rng=None
-            else:
-                wall.discard((xmin + gap, pos))
+        #
+        # WALL SAMPLING
```
What about these:

- wall creation
- wall buildup
- wall construction

Debilski replied:
*partitioning*

jbdyn replied:
I settled on "border creation" and "wall creation" as well as "partitioning" instead of "splitting" in 2e2480e7.

jbdyn committed:
Remove superfluous set creation for set-like graph node view

otizonaizit commented on pelita/maze_generator.py:
```diff
@@ -118,132 +122,165 @@ def distribute_food(all_tiles, chamber_tiles, trapped_food, total_food, rng=None
+    ngaps = max(1, ngaps)
+
+    # copy framing walls to avoid side effects
+    walls = walls.copy()
```
I don't think we have convention of favoring pure functions in Pelita. It may be a personal reference, but I personally think it is unpythonic to copy the variable at the beginning of a function, then change it in place, and then return it to then merge it with the thing you passed in to begin with. It seems to me much more pythonic to just give the function a proper name, change the thing in place and return None.

jbdyn replied:
Make it working in-place in 4e775913.

jbdyn commented:
@otizonaizit I feel like the `maze_generator` module currently suffers from encapsulation in functions which serve a module-specfic purpose, but are hardly generalized (or generalizable).

I am thinking about stuffing all of the code into `generate_maze` and then outsource smaller chunks into smaller functions.
With that, we might get rid of fractioning, naming issues and generalization blockers.

otizonaizit commented:
Hmm, as long as stuff is still testable... And that the effort does not become over engineering... Because yes, it is useful for stuff to be more general but not everything must be general for a hypothetical use case that is very unlikely to materialize.

I thought that a function that is useful to reuse while writing a bot may be `find_trapped_food` because that one you may need in order to avoid locking yourself up in a chamber while gathering. That is why I was complaining about the argument `gaps`, that is not understandable outside of `maze_generator`. 

And even if that function were reusable, I wouldn't advertise it, or it will make the whole idea of having tunnels and chambers much less challenging for the students. 

So, the code should be very readable and transparent, because we want it to be of educational grade if students look at it, but we are not inviting people to reuse it in some other random projects.

otizonaizit commented:
Hmm, as long as stuff is still testable... And that the effort does not become over engineering... Because yes, it is useful for stuff to be more general but not everything must be general for a hypothetical use case that is very unlikely to materialize.

I thought that a function that is useful to reuse while writing a bot may be `find_trapped_food` because that one you may need in order to avoid locking yourself up in a chamber while gathering. That is why I was complaining about the argument `gaps`, that is not understandable outside of `maze_generator`. 

And even if that function were reusable, I wouldn't advertise it, or it will make the whole idea of having tunnels and chambers much less challenging for the students. 

So, the code should be very readable and transparent, because we want it to be of educational grade if students look at it, but we are not inviting people to reuse it in some other random projects.

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -342,24 +420,17 @@ def generate_maze(trapped_food=10, total_food=30, width=32, height=16, rng=None)
+    # distribute food on the half maze with excluded border gaps and
+    # pacmen positions
+    chamber_tiles -= pacmen_pos
+    free_tiles = set(graph.nodes) - gaps - pacmen_pos
```
Fixed in [d696912](https://github.com/ASPP/pelita/pull/946/commits/d696912fa0ef7621543e23eda8f6319abc7dafd4).

Debilski commented on pelita/maze_generator.py:
```diff
@@ -252,53 +289,80 @@ def generate_half_maze(width, height, ngaps_center, bots_pos, rng=None):
-    # then we need to rewrite mirror to mirror a set of coordinates around the center
-    # by discarding the lower part of the border
+    # start with a full wall at the left side of the border
+    pos = width // 2 - 1
```
I find this name confusing. This seems to be something specific but calling it `pos` makes me think that it gets updated at a later point (also it is not a pos which should be 2d).

jbdyn replied:
I settled for just renaming the `pos` variable for the border position in b923ad0e.

otizonaizit commented on pelita/maze_generator.py:
that is just temporary until we decide to break numpy compatibility. This will happen soon in this very same PR. For now we keep it like this so that we are sure that refactoring does not change the generated mazes, that until now, when generated with the same seed, look exactly the same as they used to look when generated with numpy.

otizonaizit commented on pelita/maze_generator.py:
```diff
@@ -40,10 +40,18 @@
+
+# partition padding for wall position sampling
+PADDING = 2
+
```
Well, they can be added as keyword arguments to the main function, which makes it a bit uglier because then you have to pass them around. But I don't see the need to make them configurable. We haven't felt the need to change them until now.

Debilski replied:
I mean to keep them as constants here and not expect it to be user-configurable. More like as an experimentation playground (and as a tool to gain some understanding on how the maze algorithm works).

That said, `MIN_WIDTH` and `MIN_HEIGHT` work in u-v space so I am not sure these would be the right names here.

Also, we currently seem to have the minimum size for a *partition* (which is the term here for a rectangular area that still contains a wall – and that must contain a wall): So if I increase the minimum width to a high number, I can still end up with very narrow alleys because the wall could be placed at the side of a partition. [*]

Is there a reason why it is done like this? We could also have the splitting mechanism check that the walls are placed such that the resulting parts are big enough. (And if they are not then no split occurs.)

jbdyn replied:
@Debilski And what would be the meaning of a minimal partition width for you:
- the partition with walls (in line with the partition definition including surrounding walls; the minimal value allowed would be `3`) or
- the inner of the partition, i.e. the chamber size (against current partition definition, but more intuitive; minimal value allowed would be `1`)?

jbdyn replied:
I renamed the module constants in fac3897c, but left the partition definition to include the walls.

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -40,10 +40,18 @@
+PADDING = 2
+
+
 def mirror(nodes, width, height):
```
Correct. The proper term for this symmetry operation here is "inversion" with respect to a center of symmetry.

jbdyn replied:
Renamed in 6f665d6c.

I will stay with the signature to be consistent with `generate_maze`. `width` and `height` are needed separately anyways, so having `shape` as an argument would mean to have an extra pack-unpack step.

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -118,132 +122,165 @@ def distribute_food(all_tiles, chamber_tiles, trapped_food, total_food, rng=None
+        transform = identity if vertical else transposition
+
+        # map `x`-`y`-coordinates into `u`-`v`-space where the inner wall is
+        # always vertical
```
yes, or "oriented in `v`-direction".

jbdyn replied:
Rephrased in 3bdb8857.

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -118,132 +122,165 @@ def distribute_food(all_tiles, chamber_tiles, trapped_food, total_food, rng=None
+        # get the next partition
+        pmin, pmax, ngaps, vertical = partitions.pop()
+        xmin, ymin = pmin
+        xmax, ymax = pmax
```
@Debilski Or have I misunderstood your question?

jbdyn replied:
I think I got your point now.

The signature of `transform(coord_0, coord_1)` with separated variables for coordinates allows for transforming an arbitrary length of coordinates because it is just re-referencing. That's why I personally like this.
Having `transform` taking a tuple, i.e. a point, as the only argument, would also mean that I can only transform a single point at a time when I don't want to make the logic of `transform` more complicated.
However, we actually don't need that here. In fact, except for a single case, we are only transforming single points.

In this new context, I am also fine with the natural approach, so I changed the signature to accept points instead of separate coordinates in 6a4b6731.

As a bonus, it revealed minor inconsistencies in the definition of partition points.

jbdyn committed:
Properly name tile rotation function

jbdyn committed:
Add `VARIATION` constant

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -118,132 +122,165 @@ def distribute_food(all_tiles, chamber_tiles, trapped_food, total_food, rng=None
-            p += 1
+        # insert a wall only if there is some space around it in the
+        # orthogonal `u`-direction, otherwise move on with the next partition
+        if ulen < rng.randint(MIN_WIDTH, MIN_WIDTH + 2):
```
Addressed with a `VARIATION` constant in 70df282e.

jbdyn committed:
Move chamber generation from chamber tiles into test module

Currently, it is only needed for testing purposes and thereby not needed to be part of the piblic API

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -78,10 +82,10 @@ def find_trapped_tiles(graph, width, include_chambers=False):
-        chambers = []
 
-    return chamber_tiles, chambers
+        return chamber_tiles, chambers
```
Moved that block into the corresponding test module in bfbb5108.

jbdyn committed:
Use distinct phrasing for inner wall orientation

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -118,132 +122,165 @@ def distribute_food(all_tiles, chamber_tiles, trapped_food, total_food, rng=None
+        #                         wmax       pmax
+        #
+        #
+        # Note: the inner wall is always vertical.
```
Rephrased in 3bdb8857.

jbdyn committed:
Redescribe border position and length variable names

jbdyn committed:
Make transformations accept points instead of separate coordinates

jbdyn committed:
Rename binary space partitioning constants

jbdyn committed:
Fix inconsistent naming of routine for finding chamber tiles

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -55,21 +63,17 @@ def sample_nodes(nodes, k, rng=None):
 
 
-def find_trapped_tiles(graph, width, include_chambers=False):
+def find_trapped_tiles(graph, gaps, include_chambers=False):
```
I solved it in 134884ea by naming it `find_chamber_tiles`.

jbdyn committed:
Use a more descriptive argument name for border gaps

jbdyn committed:
Make binary space partitioning routing acting on walls in-place

jbdyn committed:
Rephrase code section headings

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -252,53 +289,80 @@ def generate_half_maze(width, height, ngaps_center, bots_pos, rng=None):
 
-    # Generate a wall with gaps at the border between the two homezones
+    #
+    # BORDER SAMPLING
```
Fixed in 2e2480e7.

jbdyn committed:
Make signature of binary space partitioning routine consistent

jbdyn committed:
Make ordering of node rotation consistent

jbdyn committed:
Set starting number of gaps internally

jbdyn committed:
Move all half-maze-related code into a single function

Motivation: arbitrary signatures and return values of individual functions.
Realization: logic code blocks are divided by loud comment sections.
Considerations: a `Maze` class with individual init methods.

jbdyn commented on pelita/maze_generator.py:
```diff
@@ -316,23 +380,37 @@ def generate_maze(trapped_food=10, total_food=30, width=32, height=16, rng=None)
+    # which would make it impossible to detect the main chamber
+    # assumption: border gaps are sampled centrosymmetric with always a
+    # wall segment in the middle on odd heights
+    graph.add_edges_from(edges)
```
After trying and throwing away different code changes, I found a code structure I am happy with and added cross-references in 98339855.

jbdyn committed:
Break NumPy compatibility

jbdyn committed:
Move helper function for sampling nodes

jbdyn commented:
@otizonaizit @Debilski After resolving all threads, it was time to break NumPy compatibility.

Please check, besides the code changes, whether your are still satisfied with the mazes.

There might be a slight difference because the inner partition walls are now sampled directly.
Before, we sampled gaps wider than the actual length of wall, which in turn increased the chance of less gaps as specified with `ngaps` in the inner walls and thereby more dense mazes. :thinking:

jbdyn commented:
Related comments regarding sampling and its role in NumPy compatibility:

- first try: https://github.com/ASPP/pelita/pull/856#issuecomment-2737494279
- `shuffle` vs `sample`: https://github.com/ASPP/pelita/pull/856#issuecomment-2744750789

jbdyn commented:
With the changes here (320f87a), the maze generation is still about 13% faster than without them (`main` at d2c6f147), which is in line with my earlier [comment](https://github.com/ASPP/pelita/pull/946#issuecomment-3590306194).

otizonaizit commented:
Thanks @jbdyn I really like this PR!

The maze generation code is now really readable, well structured, well tested, easy to adapt now that the parameters are not hidden but set at the beginning. The slight difference in maze density doesn't look like a problem and I am almost sure that the previous code was actually buggy, in the sense that sampling the gaps instead of the walls it was not part of the original idea but a result of the implementation using a `numpy` array.

I rebased on current `main` so that the PR is clean and easy to isolate from the rest.

I am happy to see this work finally accomplished! Thanks @jbdyn again for all the efforts, and thanks @Debilski for the review!

jbdyn commented:
Thank you, too, @otizonaizit and @Debilski!
Very happy to see this merged! :rocket:

Debilski commented on pelita/maze_generator.py:
```diff
@@ -39,49 +39,65 @@
+    if k < len(nodes):
+        return set(rng.sample(sorted(nodes), k=k))
+    else:
+        return nodes
```
Future bug potential here: We return a new set when `k < len(nodes)` but return the identical (mutable) set otherwise.

otizonaizit replied:
Could you open an issue about it? We can simply always return a new set, it's a one line fix.

jbdyn replied:
Done with #970.

Debilski commented:
Thanks for the important work. :)

Now that numpy compat is gone, I’d suggest splitting the `rng` for maze generation in two parts (three if we count foot placement, which I think should be separate for similar reasons). The idea is to use an `rng` for the gap selection that is separate from the `rng` used for the partitioning. The problem with gap selection is that it draws `k` random numbers and therefore is unstable whenever `ngaps` changes, which influences the partitioning. If they are drawn independently, then the partitioning is self-dependent (minus border constraints).

If we split the rngs, we can generate mazes that only differ by the number of initial gaps, which makes experimenting much nicer:

    ngaps = height // 4

<img width="1012" height="705" alt="image" src="https://github.com/user-attachments/assets/7a0c2602-ee7b-4730-8689-e7ae98a7e674" />

    ngaps = 2

<img width="1012" height="705" alt="image" src="https://github.com/user-attachments/assets/a75327c7-437c-4da5-ab11-32a3f85dc3af" />

    ngaps = 10

<img width="1012" height="705" alt="image" src="https://github.com/user-attachments/assets/8a18fcd3-2c22-4e8b-85d9-d14dd6d22334" />

Debilski commented:
Small side-question: This is maze `normal_034` from the old mazes. Note the long ‘tunnel’ on both sides of the bot starting positions. This seems unlikely or impossible with the current implementation of the partitioning (because we only ever partition in one dimension?). In a next iteration, we can maybe be add a small chance of generating a 2d partition.

<img width="968" height="676" alt="image" src="https://github.com/user-attachments/assets/05e1ecbe-50d1-49ea-82d0-da0628e9b978" />

otizonaizit commented:
> I’d suggest splitting the `rng` for maze generation in two parts (three if we count foot placement, which I think should be separate for similar reasons). 

Could you open an issue/feature request about this?

otizonaizit commented:
> Small side-question: This is maze `normal_034` from the old mazes. Note the long ‘tunnel’ on both sides of the bot starting positions. This seems unlikely or impossible with the current implementation of the partitioning (because we only ever partition in one dimension?).

Hmmm, but the partitioning algorithm didn't change. The only change is that we now generate walls with holes directly instead of building full walls and punching holes into them. I don't think the algorithm was doing any 2D partitioning before? @jbdyn : what do you think?

jbdyn commented:
> I’d suggest splitting the `rng` for maze generation in two parts (three if we count foot placement, which I think should be separate for similar reasons).
> 
> Could you open an issue/feature request about this?

@Debilski Or directly a PR, since you seem to already have the code?
I like the idea!

jbdyn commented:
I am not sure what the term "2D partitioning" means. :thinking:

When I recall this correctly, `normal_034` comes from the algorithm which also searched for dead ends and which has been changed in 65aa2c2e2f9f7668ac29f80e268cfd87cf6cb0ee, so `normal_034` predates the "NumPy compatible" mazes.

Nonetheless, those tunnels in `normal_034` are definitely still be possible, albeit highly unlikely, as it is such a specific configuration of walls.
